%\VignetteIndexEntry{PhenData Vignette}
%\VignetteKeywords{phenotypic data retrieval}
%\VignettePackage{PhenData}
\documentclass[a4paper]{article}

\usepackage{times}
\usepackage{a4wide}
\usepackage{url}


\SweaveOpts{keep.source=TRUE,eps=FALSE,include=FALSE,width=4,height=4.5} 

\begin{document}
\SweaveOpts{concordance=TRUE}


\title{PhenData: data retrieval from IMPC database}
\author{Jeremy Mason, Natalja Kurbatova}
\date{Modified: 08 September, 2014. Compiled: \today}

\maketitle
PhenData is an R package that allows easy access to the phenotyping data 
produced by the International Mouse 
Phenotyping Consortium (IMPC).  For more information about the IMPC project, 
please visit \url{http://www.mousephenotype.org}.   
\newline\newline
The IMPC implements a standardized set of phenotyping protocols to generate 
data. 
These standardized protocols are defined in the International Mouse Phenotyping 
Resource of Standardised Screens,
IMPReSS (\url{http://www.mousephenotype.org/impress}).  
Programmatically navigating the IMPC raw data APIs and the IMPReSS SOAP APIs 
can be technically challenging -- PhenData makes the process easier for R users. 
\newline\newline
The intended users of the PhenData package are familiar with R and would like 
easy access to the IMPC data. The data retrieved from PhenData can be directly 
used by PhenStat -- an R package that encapsulates the IMPC statistical 
pipeline, available at 
\newline
\url{http://www.bioconductor.org/packages/release/bioc/html/PhenStat.html}.

\section*{PhenData functions}
The idea of PhenData is to systematically explore the IMPC dataset's multiple 
dimensions until the correct combination of filters has been selected and then 
the data of interest downloaded.
\newline\newline
The PhenData functions can be divided into two logical groups: 
\begin{itemize}
\item "List" functions to retrieve lists of IMPC database objects and 
\item "Dataset" functions to obtain datasets from IMPC database. 
\end{itemize}
\subsection*{"List" functions}
There are "print" and "get" options for each one function in this group. 
"print" function prints the ID and name of the objects within a list. "get" 
function returns the list of IMPC IDs. There are following objects in IMPC 
database that can be obtained through appropriate functions:
\begin{itemize}
\item IMPC \textbf{phenotyping centers}: \textit{getPhenCenters}, 
\textit{printPhenCenters}.
<<R.hide, results=hide, echo=TRUE, eval=TRUE>>=
library(PhenData)
getPhenCenters()
printPhenCenters()
@
\item IMPC \textbf{pipelines}: \textit{getPipelines}, \textit{printPipelines}. 
Both functions have two arguments: 
\textit{PhenCenterName} and \textit{excludeLegacyPipelines} to exclude legacy 
pipelines from the list with default 
value set to TRUE. 
<<R.hide, results=hide, echo=TRUE, eval=TRUE>>=
library(PhenData)
getPipelines("WTSI")
printPipelines("WTSI")
@
\item IMPC \textbf{procedures} (sometimes called \textbf{screens}) that are run 
for a specified phenotyping center 
and pipeline: \textit{getProcedures}, \textit{printProcedures}. 
Both functions have two arguments: 
\textit{PhenCenterName} and \textit{PipelineID}.
<<R.hide, results=hide, echo=TRUE, eval=TRUE>>=
library(PhenData)
getProcedures("WTSI","MGP_001")
printProcedures("WTSI","MGP_001")
@
\item IMPC \textbf{parameters} that are measured within specified procedure 
for a pipeline run by phenotyping center: 
\textit{getParameters}, \textit{printParameters}. 
Functions have three arguments: \textit{PhenCenterName}, 
\textit{PipelineID} and \textit{ProcedureID}.
<<R.hide, results=hide, echo=TRUE, eval=TRUE>>=
library(PhenData)
getParameters("WTSI","MGP_001","IMPC_CBC_001")
printParameters("WTSI","MGP_001","IMPC_CBC_001")
@
\item Genetic backgrounds or, in other words, \textbf{strains} from which 
the knockout mice were derived for a specific combination of pipeline, procedure
and parameter for a phenotyping center: \textit{getStrains}, 
\textit{printStrains}.
 Strain's ID is MGI (\url{http://www.informatics.jax.org/}) ID or temporary ID 
 if the MGI is not assigned yet.  
There are following arguments for both functions: \textit{PhenCenterName},
\textit{PipelineID}, \textit{ProcedureID} 
and \textit{ParameterID}.
<<R.hide, results=hide, echo=TRUE, eval=TRUE>>=
library(PhenData)
getStrains("WTSI","MGP_001","IMPC_CBC_001","IMPC_CBC_003_001")
printStrains("WTSI","MGP_001","IMPC_CBC_001","IMPC_CBC_003_001")
@
\item \textbf{Genes} that are reported for a specified combination of parameter,
procedure, pipeline and phenotyping 
center: \textit{getGenes}, \textit{printGenes}. 
Gene's ID is MGI (\url{http://www.informatics.jax.org/}) ID or 
temporary ID if the MGI is not assigned yet. 
There are following arguments: \textit{PhenCenterName}, 
\textit{PipelineID}, \textit{ProcedureID}, \textit{ParameterID}, 
\textit{StrainID} is an optional argument.
<<R.hide, results=hide, echo=TRUE, eval=TRUE>>=
library(PhenData)
getGenes("WTSI","MGP_001","IMPC_CBC_001","IMPC_CBC_003_001")
printGenes("WTSI","MGP_001","IMPC_CBC_001",
"IMPC_CBC_003_001","MGI:2159965")
@
\item \textbf{Alleles} that are processed for a specified combination of 
parameter, procedure, pipeline and 
phenotyping center: \textit{getAlleles}, \textit{printAlleles}. Allele's ID 
is MGI (\url{http://www.informatics.jax.org/}) ID or temporary ID 
if the MGI is not assigned yet. 
There are following arguments: \textit{PhenCenterName}, \textit{PipelineID}, 
\textit{ProcedureID}, 
\textit{ParameterID} and \textit{StrainID}, which is an optional argument.
<<R.hide, results=hide, echo=TRUE, eval=TRUE>>=
library(PhenData)
getAlleles("WTSI","MGP_001","IMPC_CBC_001",
"IMPC_CBC_003_001","MGI:5446362")
printAlleles("WTSI","MGP_001","IMPC_CBC_001","IMPC_CBC_003_001")
@
\item \textbf{Zygosities} (homozygous, heterozygous and hemizygous) for 
mice that were measured for a gene/allele 
for a specified combination of parameter, procedure, pipeline and 
phenotyping center: \textit{getZygosities}. 
There are following arguments of the \textit{getZygosities} function:  
\textit{PhenCenterName}, \textit{PipelineID}, 
\textit{ProcedureID}, \textit{ParameterID}, \textit{StrainID}, which is an 
optional argument, \textit{GeneID}, 
which is an optional argument and finally \textit{AlleleID}, 
which  is also an optional argument. 
<<R.hide, results=hide, echo=TRUE, eval=TRUE>>=
library(PhenData)
getZygosities("WTSI","MGP_001","IMPC_CBC_001","IMPC_CBC_003_001",
StrainID="MGI:5446362",AlleleID="EUROALL:64")
@
\end{itemize}


\subsection*{"Dataset" functions}
The PhenData package "dataset" functions allow extraction of IMPC datasets 
where potential phenodeviants data and 
control data are matched together according to the internal IMPC rules. 
There are two functions in this group:  
\textit{getIMPCTable} and  \textit{getIMPCDataset}.
\newline\newline
Function \textit{getIMPCTable} creates a table with combination of objects 
to define IMPC datasets according to 
the parameters that have been passed to the function. Table is stored using 
comma separated format 
in the file specified by user.  Function's arguments are:
\begin{itemize}
\item \textit{fileName} -- defines name of the file where to save resulting 
table with IMPC objects, 
mandatory argument with default value set to "PhenData\_IMPC"; 
\item \textit{PhenCenterName} -- IMPC phenotyping center; 
\item \textit{PipelineID} -- IMPC pipeline ID;
\item \textit{ProcedureID} -- IMPC procedure ID;
\item \textit{ParameterID} -- IMPC parameter ID;
\item \textit{AlleleID} -- allele ID;
\item \textit{StrainID} -- strain ID;
\item \textit{multipleFiles} -- flag: "FALSE" value to get all records into 
one specified file; "TRUE" value (default) 
to split records across multiple files named starting with "fileName";
\item \textit{recordsPerFile} -- number that specifies how many records 
to write into one file with default value set to 10000.
\end{itemize}
Example of usage: 
<<R.hide, results=hide, echo=TRUE, eval=FALSE>>=
library(PhenData)
getIMPCTable("./IMPCData.csv","WTSI","MGP_001","IMPC_CBC_001")  
@
All possible combinations are stored now into the file "IMPCData.csv" 
in comma separated format. 
There are six columns in the saved file: "Phenotyping Center", "Pipeline", 
"Screen/Procedure", "Parameter", 
"Allele" and "Function to get IMPC Dataset". 
The last one column "Function to get IMPC Dataset" contains the prepared R 
code to call the function 
\textit{getIMPCDataset} with appropriate parameters and to obtain the dataset.
\newline\newline
For example, take the first row and the last column with dataset function call: 
<<R.hide, results=hide, echo=TRUE, eval=FALSE>>=
library(PhenData)
IMPC_dataset1 <- getIMPCDataset("WTSI","MGP_001","IMPC_CBC_001",
"IMPC_CBC_003_001","MGI:4431644") 
@
Now IMPC dataset is obtained and ready to use with PhenStat, for example:
<<R.hide, results=hide, echo=TRUE, eval=FALSE>>=
library(PhenStat)
testIMPC1 <- PhenList(dataset=IMPC_dataset1,
testGenotype="MDTZ",
refGenotype="+/+",
dataset.colname.genotype="Colony") 
@
\end{document}